oidc/refresh_token.go

```
@@ -0,0 +1,82 @@
+	ExpiresIn        string `json:"expires_in"`
+	IDToken          string `json:"id_token"`
+	Scope            string `json:"scope"`
+	TokenType        string `json:"token_type"`
+	Error            string `json:"error"`
+	ErrorDescription string `json:"error_description"`
+}
+
+// RefreshRequest represents the request that the token refresh endpoint expects
+// in the JSON body. It gets marshalled to JSON.
+type RefreshRequest struct {
+	ClientID     string `json:"client_id"`
+	ClientSecret string `json:"client_secret"`
+	GrantType    string `json:"grant_type"`
+	RefreshToken string `json:"refresh_token"`
+}
+
+// RefreshToken performs a POST call to the auth0 token endpoint with a refresh token
+// and returns a RefreshToken response, which includes a new access token.
+func RefreshToken(refreshToken string) (refreshResponse RefreshResponse, err error) {
+	payload := strings.NewReader(fmt.Sprintf(`{
+    "grant_type":"refresh_token",
+    "client_id": "%s",
+    "refresh_token": "%s"
+  }`, clientID, refreshToken))
+
+	req, err := http.NewRequest("POST", tokenURL, payload)
+	if err != nil {
+		refreshResponse.Error = "unknown_error"
+		refreshResponse.ErrorDescription = "Unable to construct POST request for Auth0."
+		return refreshResponse, microerror.Maskf(authorizationError, refreshResponse.Error)
+	}
+
+	req.Header.Add("content-type", "application/json")
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		refreshResponse.Error = "unknown_error"
+		refreshResponse.ErrorDescription = "Unable to perform POST request to Auth0."
+		return refreshResponse, microerror.Maskf(authorizationError, refreshResponse.Error)
+	}
+
+	defer res.Body.Close()
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		refreshResponse.Error = "unknown_error"
+		refreshResponse.ErrorDescription = "Got an unparseable error from Auth0. Possibly the Auth0 service is down. Try again later."
+		return refreshResponse, microerror.Maskf(authorizationError, refreshResponse.Error)
```

We should return a specific error here, not `authorizationError`. In `handleCommonError` for `authorizationError` we print `Unauthorized` as a headline, which can be misleading. In this case here the authentication process failed, so we should allow ourselves to show a more appropriate message to the user.